custom/shared/hooks/useSharedState.js

```diff
@@ -0,0 +1,111 @@
+import React, { useState, useEffect, useCallback, useRef } from 'react';
+import { useCallState } from '../contexts/CallProvider';
+
+// @params initialValues - initial values of the shared state.
+// @params broadcast - share the state with the other participants whenever the state changes.
+export const useSharedState = ({ initialValues = {}, broadcast = true }) => {
+  const { callObject } = useCallState();
+  const stateRef = useRef(null);
+
+  const [state, setState] = useState({
+    sharedState: initialValues,
+    setAt: undefined,
+  });
+
+  // handling the app-message event, to check if the state is being shared.
+  const handleAppMessage = useCallback(
+    event => {
+      // two types of events -
+      // 1. Request shared state (request-shared-state)
+      // 2. Set shared state (set-shared-state)
+      switch (event.data?.message?.type) {
+        // if we receive a request-shared-state message type, we check if the user has any previous state,
+        // if yes, we will send the shared-state to everyone in the call.
+        case 'request-shared-state':
+          if (!stateRef.current.setAt) return;
+          callObject.sendAppMessage(
+            {
+              message: {
+                type: 'set-shared-state',
+                value: stateRef.current,
+              },
+            },
+            '*',
+          );
+          break;
+        // if we receive a set-shared-state message type then, we check the state timestamp with the local one and
+        // we set the latest shared-state values into the local state.
+        case 'set-shared-state':
+          if (
+            stateRef.current.setAt &&
+            new Date(stateRef.current.setAt) > new Date(event.data.message.value.setAt)
+          )
+            return;
+          setState(event.data.message.value);
+          break;
+      }
+    },
+    [stateRef, callObject],
+  );
+
+  // whenever local user joins, we request the state from everyone in the call.
+  const handleJoinedMeeting = useCallback(() => {
+    const randomDelay = 1000 + Math.ceil(1000 * Math.random());
+    setTimeout(() => {
+      const callObjectParticipants = callObject.participants();
+      const participants = Object.values(callObjectParticipants);
+      const localParticipant = callObjectParticipants.local;
+      if (participants.length > 1) {
+        const remoteParticipants = participants.filter((p) => !p.local && new Date(p.joined_at) < new Date(localParticipant.joined_at));
+        const randomPeer = remoteParticipants[Math.floor(Math.random() * (remoteParticipants.length - 1))];
+        callObject.sendAppMessage(
+          {
+            message: {
+              type: 'request-shared-state',
+            },
+          },
+          randomPeer.user_id,
+        );
+      }
+    }, randomDelay);
```

Comment:
A few more suggestions:

- It's possible, that a remote peer is requested for state, who doesn't have any state to share, yet (because they're awaiting to receive the shared state themselves). This can mean that the `request-shared-state` message is not responded. Changing the `setTimeout` to a `setInterval` makes sure that in that case the request is repeated (potentially to another peer, but that's fine).
- We should store the `interval` in a ref (`const requestIntervalRef = useRef(null);`) and cancel the interval when receiving a `set-shared-state` message. This has the benefit that 1 request can potentially fulfill multiple other newly joined peers. The `randomDelay` also makes more sense with that change.
- We should also clear the interval in the `useEffect's` return hook, for good measure ☺️